### DISCUSSION TITLE
SeleniumBase in Docker

### DISCUSSION THREAD

**Discussion by pol4xer** [seleniumbase/SeleniumBase#2619] [General]
Hello! 
I am trying to use **SeleniumBase** in Docker container.
I am building it in **Mac M1**, for image I use **python:3.8-bullseye**.
When I try to do following:
```python3
from seleniumbase import Driver
driver = Driver(headless=True)
```
I am getting errors: `selenium.common.exceptions.WebDriverException: Message: Unsupported platform/architecture combination: linux/aarch64` and `selenium.common.exceptions.NoSuchDriverException: Message: Unable to obtain driver for chrome using Selenium Manager.`

I went to google, tried some things, and eventually I succeeded for WebDriver:
```python3
from selenium import webdriver
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.chrome.options import Options
options = Options()
options.add_argument("--headless")
options.add_argument("start-maximized")
options.add_argument("disable-infobars")
options.add_argument("--disable-extensions")
options.add_argument("--disable-gpu")
options.add_argument("--disable-dev-shm-usage")
options.add_argument("--no-sandbox")
webdriver.Chrome(service=Service("/usr/bin/chromedriver"), options=options)
```
And code with `webdriver` returned: `<selenium.webdriver.chrome.webdriver.WebDriver (session="4e1a81ac8ab545a888b49cbcfdc11109")>`

What should I do, to be able to use `Driver` from SeleniumBase?
**P.S.** I don't want to change image or download already built image. I want to understand how it works and what I do wrong.
**P.P.S**
- I tried to add `/usr/bin/chromedriver` to Driver: ```Driver(headless=True, binary_location='/usr/bin/chromedriver')``` and it didn't work.
- ChromeDriver I installed with `apt-get install chromium-driver`, because locally I develop on Mac, but my server on Ubuntu. So this command allows me to install on mac for mac, on server for server, with no specific versions specified

Thanks in advance!

**Comment by mdmintz** [comment]
The `Unsupported platform/architecture combination: linux/aarch64` has some info here:
https://groups.google.com/g/selenium-users/c/MRfrR7pu_8I
with an answer here: https://groups.google.com/g/selenium-users/c/MRfrR7pu_8I/m/F9U3eMC0AAAJ

SeleniumBase also has a sample `Dockerfile` here: https://github.com/seleniumbase/SeleniumBase/blob/master/Dockerfile

**Reply by pol4xer** [reply]
I marked, that with `webdriver` it works: `And code with webdriver returned: <selenium.webdriver.chrome.webdriver.WebDriver (session="4e1a81ac8ab545a888b49cbcfdc11109")>`

I can't use `Driver` from seleniumbase. Do you know what should I write to be able to run `Driver` in Docker? 

**P.S.** I also tried this: ```Driver(headless=True,headless2=True,chromium_arg='--headless,start-maximized,disable-infobars,--disable-extensions,--disable-gpu,--disable-dev-shm-usage,--no-sandbox')``` - The same error, despite that `webdriver` with these arguments worked well

**Reply by mdmintz** [reply]
`binary_location` is for the location of Chrome (not `chromedriver`). To use a specific driver, place it into the `seleniumbase/drivers` folder.

**Reply by pol4xer** [reply]
Okay, I tried to move `chromdriver` to `seleniumbase/drivers/` and it didn't help.
So I have this: 
- Code with `webdriver` is working.
- Code with `Driver` doesn't work.
- Arguments are the same
- `chromedriver` is the same

Where is the problem? What am I doing incorrectly?

**Reply by mdmintz** [reply]
> **P.S.** I also tried this: ```Driver(headless=True,headless2=True,chromium_arg='--headless,start-maximized,disable-infobars,--disable-extensions,--disable-gpu,--disable-dev-shm-usage,--no-sandbox')``` - The same error, despite that `webdriver` with these arguments worked well

There are multiple issues with this line. One, you can't use both headless modes at the same time. There's the regular `headless` mode, and there's also the newer one (`--headless=new` via Chrome Options). Two, to maximize the browser window, call `driver.maximize_window()` after the browser has been launched. Three, several of those options are already being set, eg. https://github.com/seleniumbase/SeleniumBase/blob/e4abd8e8beec5a85b31fcddcbdfd231812af15c4/seleniumbase/core/browser_launcher.py#L1049 which means that if you try setting it twice, something might not work. Check to see what is already being set before adding the option into `chromium_arg`.

Another thing you mentioned is:

> "Unable to obtain driver for chrome using Selenium Manager"

Selenium Manager isn't part of SeleniumBase. If you're seeing that, I need to know why you saw it. You may need to debug with `pdb` to reach the exact line.
